corehq/messaging/smsbackends/icds/models.py

```diff
@@ -0,0 +1,132 @@
+from urllib import urlencode
+from urllib2 import urlopen
+
+from crispy_forms import layout as crispy
+from django.conf import settings
+from django.utils.translation import ugettext_lazy as _
+
+from corehq.apps.sms.models import SQLSMSBackend, SMS
+from corehq.apps.sms.forms import BackendForm
+from corehq.apps.sms.util import strip_plus
+from dimagi.utils.django.fields import TrimmedCharField
+from dimagi.utils.logging import notify_exception
+
+ERROR_CODES = {
+    "-2": "Invalid credentials",
+    "-3": "Empty mobile number",
+    "-4": "Empty message",
+    "-5": "HTTPS disabled",
+    "-6": "HTTP disabled",
+    "-13": "Internal Error",
+    "-201": "Email Delivery Disabled",
+    "-401": "Invalid Scheduled Time",
+    "-404": "Invalid MsgType",
+    "-406": "Invalid Port",
+    "-407": "Invalid Expiry minutes",
+    "-408": "Invalid Customer Reference Id",
+    "-409": "Invalid Bill Reference Id",
+    "-410": "Invalid Destination Address",
+    "-432": "Invalid Bill Reference Id Length",
+    "-433": "Invalid Customer Reference Id Length",
+}
+
+RETRY_ERROR_CODES = {"-13"}
+
+
+class ICDSException(Exception):
+    pass
+
+
+class ICDSBackendForm(BackendForm):
+    username = TrimmedCharField(
+        label=_('Username'),
+    )
+    pin = TrimmedCharField(
+        label=_('PIN'),
+    )
+    sender_id = TrimmedCharField(
+        label=_('Sender ID'),
+    )
+
+    @property
+    def gateway_specific_fields(self):
+        return crispy.Fieldset(
+            _("ICDS Settings"),
+            'username',
+            'pin',
+            'sender_id',
+        )
+
+
+class SQLICDSBackend(SQLSMSBackend):
+
+    class Meta:
+        app_label = 'sms'
+        proxy = True
+
+    @classmethod
+    def get_available_extra_fields(cls):
+        return [
+            'username',
+            'pin',
+            'sender_id',
+        ]
+
+    @classmethod
+    def get_api_id(cls):
+        return 'ICDS'
+
+    @classmethod
+    def get_generic_name(cls):
+        return "ICDS"
+
+    @classmethod
+    def get_form_class(cls):
+        return ICDSBackendForm
+
+    def get_response_code(self, response):
+        api_code_string = "~code=API"
+        begin_response = response.find(api_code_string)
+        if begin_response == -1:
+            return None
+        begin_response += len(api_code_string)
+        end_response = response[begin_response:].find('&')
+        if end_response == -1:
+            return None
+        return response[begin_response:end_response]
```

Comment:
we should also trim whitespace from the final result since it looks like in some of the examples there can be whitespace around it